BUG: Fix Series.nlargest for integer boundary values

jschendel opened the pull request:
- [X] closes #21426
- [X] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

Also added some similar tests for float and datetimelike dtypes to ensure that the behavior is as desired.

jschendel committed:
BUG: Fix Series.nlargest for integer boundary values

gfyoung commented on pandas/tests/series/test_analytics.py:
```diff
@@ -2028,6 +2030,33 @@ def test_n(self, n):
+        result = getattr(s, method)(3)
+        expected_idxr = [0, 1, 2] if method == 'nsmallest' else [3, 2, 1]
+        expected = s.loc[expected_idxr]
+        tm.assert_series_equal(result, expected)
```
I like the thought of using parameterization, but this is a little hard to read.  I would recommend the following:

* Split this test into three tests based on which conditional it hits.
* Extract lines 2054-2058 as a separate (non-test) method, where you pass in `vals` as a parameter i.e.

~~~python
def _check_nselect(self, vals, method):
   s = Series(...)
   ...
   tm.assert_series_equal(result, expected)
~~~

jschendel replied:
done

gfyoung commented:
> No whatsnew entry for the time being since this was tagged as 0.23.2, and no whatsnew has been created for that release yet.

@jschendel : Not a problem.  Just wait for #21433 and rebase.

gfyoung commented:
#21433 has been merged.  Feel free to rebase.

jschendel committed:
add whatsnew

jschendel committed:
split tests

gfyoung reviewed (approved):
Nice!

cc @jreback

jschendel committed:
review edits

gfyoung commented on pandas/tests/series/test_analytics.py:
```diff
@@ -2028,6 +2032,40 @@ def test_n(self, n):
+        expected = s.loc[expected_idxr]
+        tm.assert_series_equal(result, expected)
+
+    @pytest.mark.parametrize('dtype', [
```
@jschendel brought up a good point in https://github.com/pandas-dev/pandas/pull/21456#discussion_r195266459 that we might have some intersection with this PR and #21456.  How best to work that out?

jschendel replied:
FWIW I've modified the fixtures I wrote to exactly match @gfyoung's (aside from removing float16 and fixing the `ALL_INT_DTYPES` param).  Were basically identical anyways.  I've got all my review related changes ready, so can push whenever.

jschendel commented:
Addressed review comments.  Copied @gfyoung's dtype fixtures.  Did some additional cleaning of `frame/test_analytics.py.` to use one of the new fixtures I created, and to use multiple instances of `parametrize` instead of jamming everything into a single one with `product`.

jreback commented:
thanks @jschendel

jreback commented:
can you create an issue to use these fixtures project wide?

jschendel commented:
> can you create an issue to use these fixtures project wide?

Created #21500 for this
